### PR TITLE
feat:[SRM-15553]:Implemented Least performance Request Based Composit…

### PR DESCRIPTION
**Pull request by shash-harness**
…e SLO

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52987)

### CODE REVIEW

**Comment by shash-harness** [comment]
trigger uts

**Comment by shash-harness** [comment]
trigger uts

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/beans/slospec/WeightedAverageCompositeSLOEvaluator.java [hunk 1]
```diff
@@ -7,12 +7,37 @@
...
+
+  @Override
+  public Double getSloPercentage(CompositeServiceLevelObjective compositeServiceLevelObjective,
+      CompositeSLORecord sloRecord, CompositeSLORecord prevSLORecord) {
+    double sloPercentage = 0.0;
+    for (CompositeServiceLevelObjective.ServiceLevelObjectivesDetail serviceLevelObjectivesDetail :
+        compositeServiceLevelObjective.getServiceLevelObjectivesDetails()) {
+      Double weightage = serviceLevelObjectivesDetail.getWeightagePercentage() / 100;
+      SLIRecord sliRecord = sloRecord.getScopedIdentifierSLIRecordMap().get(
+          serviceLevelObjectiveV2Service.getScopedIdentifier(serviceLevelObjectivesDetail));
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/beans/slospec/WeightedAverageCompositeSLOEvaluator.java:36]: don't use serviceClass, use ScopedInformation.

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/beans/slospec/LeastPerformanceCompositeSLOEvaluator.java [hunk 2]
```diff
@@ -7,13 +7,37 @@
...
+    double sloPercentage = 0.0;
+    for (CompositeServiceLevelObjective.ServiceLevelObjectivesDetail serviceLevelObjectivesDetail :
+        compositeServiceLevelObjective.getServiceLevelObjectivesDetails()) {
+      Double weightage = serviceLevelObjectivesDetail.getWeightagePercentage() / 100;
+      SLIRecord sliRecord = sloRecord.getScopedIdentifierSLIRecordMap().get(
+          serviceLevelObjectiveV2Service.getScopedIdentifier(serviceLevelObjectivesDetail));
+      SLIRecord prevSLIRecord = prevSLORecord.getScopedIdentifierSLIRecordMap().get(
+          serviceLevelObjectiveV2Service.getScopedIdentifier(serviceLevelObjectivesDetail));
+      sloPercentage = Math.max(
+          weightage * (SLIValue.getRunningCountDifference(sliRecord, prevSLIRecord).sliPercentage()), sloPercentage);
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/beans/slospec/LeastPerformanceCompositeSLOEvaluator.java:39]: You are taking max of sliPercentage, sliPercentage is good/total. We want max badCount. you need Min here not Max.

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/services/impl/GraphDataServiceImpl.java [hunk 3]
```diff
@@ -188,16 +193,9 @@ private double getSLOValue(SLOValue sloValue, CompositeServiceLevelObjective com
...
-    for (CompositeServiceLevelObjective.ServiceLevelObjectivesDetail serviceLevelObjectivesDetail :
-        compositeServiceLevelObjective.getServiceLevelObjectivesDetails()) {
-      Double weightage = serviceLevelObjectivesDetail.getWeightagePercentage() / 100;
-      SLIRecord sliRecord = sloRecord.getScopedIdentifierSLIRecordMap().get(
-          serviceLevelObjectiveV2Service.getScopedIdentifier(serviceLevelObjectivesDetail));
-      SLIRecord prevSLIRecord = prevSLORecord.getScopedIdentifierSLIRecordMap().get(
-          serviceLevelObjectiveV2Service.getScopedIdentifier(serviceLevelObjectivesDetail));
-      sloPercentage += weightage * (SLIValue.getRunningCountDifference(sliRecord, prevSLIRecord).sliPercentage());
-    }
+    double sloPercentage =
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/services/impl/GraphDataServiceImpl.java:196]: There will be changes in getBudgetBurnDown as well

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/beans/slospec/CompositeSLOEvaluator.java [hunk 4]
```diff
@@ -7,12 +7,18 @@
...
 import java.util.ArrayList;
 import java.util.List;
 
 public abstract class CompositeSLOEvaluator {
   public abstract Double evaluate(List<Double> weightage, List<Integer> sliValues);
 
+  public abstract Double getSloPercentage(CompositeServiceLevelObjective compositeServiceLevelObjective,
+      CompositeSLORecord sloRecord, CompositeSLORecord prevSLORecord);
+
   List<Double> getSLOValuesOfIndividualSLIs(List<Double> weightage, List<Integer> sliValues) {
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/servicelevelobjective/beans/slospec/CompositeSLOEvaluator.java:22]: add Uts after fixing

**Review by Varsha-lalwani — CHANGES_REQUESTED** [review]

**Comment by shash-harness** [comment]
trigger uts

**Comment by shash-harness** [comment]
trigger uts

**Comment by shash-harness** [comment]
trigger uts

**Comment by shash-harness** [comment]
trigger uts

**Comment by shash-harness** [comment]
trigger ut9

**Comment by shash-harness** [comment]
trigger messagecheck

**Comment by shash-harness** [comment]
trigger uts

**Comment by shash-harness** [comment]
trigger messagecheck

**Comment by shash-harness** [comment]
trigger ut3

**Comment by shash-harness** [comment]
trigger uts

**Pull request closed without merging**